File: c_files/raylib/rader_sweep.c
#include "raylib.h"
#include <math.h>

typedef struct Ring
{
    Vector2 center;
    float inner_radius;
    float outer_radius;
    float start_angle;
    float end_angle;
    int segments;
    Color color;

} Ring;

typedef struct Line
{
    Vector2 start_pos;
    Vector2 end_pos;
    float thickness;
    Color color;

} Line;

int main(void)
{
    const int SCREEN_WIDTH = 800;
    const int SCREEN_HEIGHT = 600;
    const char SCREEN_TITLE[] = "Radar Sweep";
    const Color SCREEN_BACKGROUND = SKYBLUE;

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    Ring ring;
    ring.center = (Vector2){(float)GetScreenWidth() / 2,
                            (float)GetScreenHeight() / 2};
    ring.inner_radius = 250.0f;
    ring.outer_radius = ring.inner_radius + 15;
    ring.start_angle = 0.0f;
    ring.end_angle = 360.0f;
    ring.segments = 32;
    ring.color = DARKGREEN;

    Line line;
    line.start_pos = ring.center;
    line.end_pos = (Vector2){0, 0};
    line.thickness = 10.0f;
    line.color = YELLOW;

    float angle = 0.0f;
    float sweep_speed = 50.0f;

    while (!WindowShouldClose())
    {
        angle += sweep_speed * GetFrameTime();
        if (angle > 360)
        {
            angle = 0.0f;
        }

        line.end_pos.x = ring.center.x + cosf(angle * DEG2RAD) * ring.inner_radius;
        line.end_pos.y = ring.center.y + sinf(angle * DEG2RAD) * ring.inner_radius;

        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        DrawRing(ring.center,
                 ring.inner_radius,
                 ring.outer_radius,
                 ring.start_angle,
                 ring.end_angle,
                 ring.segments,
                 ring.color);

        DrawLineEx(line.start_pos,
                   line.end_pos,
                   line.thickness,
                   line.color);

        EndDrawing();
    }

    CloseWindow();

    return 0;
}

File: python/raylib/rader_sweep.py
import pyray as p
from math import radians, cos, sin

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Radar Sweep"
SCREEN_BACKGROUND = p.SKYBLUE


class Ring:
    def __init__(self):
        self.center = p.Vector2(p.get_screen_width() / 2, p.get_screen_height() / 2)
        self.inner_radius = 250
        self.outer_radius = 260
        self.start_angle = 0
        self.end_angle = 360
        self.segments = 64
        self.color = p.DARKGREEN

    def draw(self):
        p.draw_ring(
            self.center,
            self.inner_radius,
            self.outer_radius,
            self.start_angle,
            self.end_angle,
            self.segments,
            self.color,
        )


class Line:
    def __init__(self, center_x, center_y, inner_radius):
        self.center_x = center_x
        self.center_y = center_y
        self.inner_radius = inner_radius
        self.start = p.Vector2(center_x, center_y)
        self.end = p.Vector2(0, 0)
        self.thickness = 10
        self.color = p.YELLOW
        self.angle = 0
        self.speed = 50

    def draw(self):
        p.draw_line_ex(self.start, self.end, self.thickness, self.color)

    def update(self):
        self.angle += self.speed * p.get_frame_time()
        if self.angle >= 360:
            self.angle = 0

        self.end.x = self.center_x + cos(radians(self.angle)) * self.inner_radius
        self.end.y = self.center_y + sin(radians(self.angle)) * self.inner_radius


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    ring = Ring()
    line = Line(ring.center.x, ring.center.y, ring.inner_radius)

    while not p.window_should_close():
        line.update()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        ring.draw()
        line.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
